### ISSUE TITLE
`yarn global upgrade` doesn't upgrade packages

### ISSUE THREAD

**Issue by mgol** [yarnpkg/yarn#5001]
**Do you want to request a *feature* or report a *bug*?**

A bug.

**What is the current behavior?**

`yarn global upgrade` doesn't upgrade packages.

**If the current behavior is a bug, please provide the steps to reproduce.**

First, invoke the following for some package named `some-package`:
```
yarn global add some-package
```
Then wait for `some-package` to get an update. Afterwards, run:
```
yarn global upgrade
```
yarn will print a lot of package names suggesting it did upgrade something but, in fact, `some-package` will stay at its previous version.

**What is the expected behavior?**

The command should upgrade all the packages to their `latest` versions. In its current state the command doesn't seem to be doing anything useful.

**Please mention your node.js, yarn and operating system version.**
Node 8.9.1, Yarn 1.3.2, macOS 10.13.1 (17B48)

**Comment by rally25rs** [comment]
Could you provide an example package and versions? What is the output of `yarn global outdated`?

Keep in mind that `upgrade` respects semver range specified in package.json. I believe the default location on OSX is `~/.config/yarn/global/package.json`

So `yarn global add some-package` would add a carret-range, something like `^1.0.0` so if a `2.0.0` comes out, `yarn global upgrade` would not upgrade to that because it doesn't fit the range. You would specify the `--latest/-L` flag to ignore the semver range and get the `latest` as tagged in the registry.

**Comment by mgol** [comment]
OK, that explains a lot as I have `save-prefix` set to `''` in settings.

I think this behavior is surprising nonetheless. Global packages differ from local ones in that they cannot be versioned carefully for a specific project, they're shared for all the code run from the current machine. Therefore, semver & breaking changes are less important.

It'd be useful to have a command that upgrades all currently installed global packages to their latest versions as npm allows.

**Comment by rally25rs** [comment]
> Therefore, semver & breaking changes are less important.

Yeah, I don't disagree with you there. Internally `yarn global` is really calling the "normal" `upgrade` command, and the behavior and flags keep it consistent with local/project `upgrade` but it is a little weird in the "global" scope to not just always get the latest. But it does make it so that `global upgrade`, `upgrade`, and `upgrade-interactive` all have the same behavior with and without the `--latest` flag.

> It'd be useful to have a command that upgrades all currently installed global packages to their latest versions as npm allows.

`yarn global upgrade -L` would do that.

**Comment by rally25rs** [comment]
Also I think you could make a global `.yarnrc` that contains:

```
--global.latest true
```

to basically auto-pass the `--latest` flag by default, then `yarn global upgrade` would act like you had passed the `--latest` flag.

**Comment by mgol** [comment]
> But it does make it so that `global upgrade`, `upgrade`, and `upgrade-interactive` all have the same behavior with and without the `--latest` flag.

I understand the appeal logic-wise. One more argument for a default `--latest` for `global upgrade` - in regular projects people look at the `package.json`. Sometimes they update a dep there manually, sometimes via Yarn & then add to version control (together with `yarn.lock`); in both cases it's visible to the developer, though. For global installs the config seems to be a side-effect of how global packages are handled internally; `yarn global list` doesn't list version ranges of the global packages and the file itself is hidden under `~/.config/yarn/global/package.json` where most people wouldn't think to look.

>> It'd be useful to have a command that upgrades all currently installed global packages to their latest versions as npm allows.
>>
> `yarn global upgrade -L` would do that.

I've tried `yarn global upgrade --latest` but it doesn't upgrade my packages either. That's my `yarn` config if that matters:
```js
{ 'version-tag-prefix': 'v',
  'version-git-tag': true,
  'version-git-sign': false,
  'version-git-message': 'v%s',
  'init-version': '1.0.0',
  'init-license': 'MIT',
  'save-prefix': '',
  'ignore-scripts': false,
  'ignore-optional': false,
  registry: 'https://registry.yarnpkg.com',
  'strict-ssl': true,
  'user-agent': 'yarn/1.3.2 npm/? node/v8.9.1 darwin x64',
  lastUpdateCheck: 1511823359616,
  prefix: '/usr/local/yarn/' }
```

**Comment by sharno** [comment]
Same here, `yarn global upgrade --latest` doesn't upgrade to latest packages. Have to run `yarn global add flow-bin@latest` to upgrade for example

**Comment by zzswang** [comment]
+1

**Comment by tedmiston** [comment]
It's confusing that Yarn handles upgrades differently than every other package manager.

This is particularly confusing with globally installed packages like CLIs where I just want the latest version installed always and am not concerned about breaking dependencies.

I've run into this issue with Yarn multiple times upgrading [Zeit Now](https://github.com/zeit/now-cli) as they increase their major version regularly — TODO times (TODO, ..., 13) in the past year.

Running `yarn global upgrade now` prints a bunch of output as if the package were upgraded.  It currently does not print any sort of warning that the package was not actually upgraded.  Running `yarn global add now` does successfully achieve what I expected upgrade to do.

I understand why you apply this constraint, but not installing the next major version based on an assumed concern with semver is also kind of arbitrary at best and confusing at worst to someone in my use case (where major version is irrelevant), so it would be nice to at least show a warning.

**Comment by seminelee** [comment]
same here

**Comment by DemonCloud** [comment]
same bug
